cmd/state.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/pulumi/pulumi/pkg/util/contract"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/apitype"
+	"github.com/pulumi/pulumi/pkg/backend/display"
+	"github.com/pulumi/pulumi/pkg/diag/colors"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/resource/edit"
+	"github.com/pulumi/pulumi/pkg/resource/stack"
+	"github.com/pulumi/pulumi/pkg/util/cmdutil"
+	"github.com/spf13/cobra"
+	survey "gopkg.in/AlecAivazis/survey.v1"
+	surveycore "gopkg.in/AlecAivazis/survey.v1/core"
+)
+
+func newStateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "state",
+		Short: "Edit the current stack's state",
+		Long: `Edit the current stack's state
+		
+Subcommands of this command can be used to surgically edit parts of a stack's state. These can be useful when
+troubleshooting a stack or when performing specific edits that otherwise would require editing the state file by hand.`,
+		Args: cmdutil.NoArgs,
+	}
+
+	cmd.AddCommand(newStateDeleteCommand())
+	cmd.AddCommand(newStateUnprotectCommand())
+	return cmd
+}
+
+// locateStackResource attempts to find a unique resource associated with the given URN in the given snapshot. If the
+// given URN is ambiguous and this is an interactive terminal, it prompts the user to select one of the resources in
+// the list of resources with identical URNs to operate upon.
+func locateStackResource(opts display.Options, snap *deploy.Snapshot, urn resource.URN) (*resource.State, error) {
+	candidateResources := edit.LocateResource(snap, urn)
+	switch {
+	case len(candidateResources) == 0: // resource was not found
+		return nil, nil
+	case len(candidateResources) == 1: // resource was unambiguously found
+		return candidateResources[0], nil
+	}
+
+	// If there exist multiple resources that have the requested URN, prompt the user to select one if we're running
+	// interactively. If we're not, early exit.
+	if !cmdutil.Interactive() {
+		return nil, errors.Errorf("Resource URN %q ambiguously refers to multiple resources", urn)
```

Comment:
nit: consider printing out the different urns.  also, since urn is long it might be better to say:

```
Resource URN ambiguously refered to multiple resources.  Did you mean:\n
  urn1
  urn2
  urn3
...
```